diffrax/_adjoint.py

```diff
@@ -852,3 +865,427 @@ def loop(
+            y0,
+            (original_solver_state, z0),
+            grad_y0,
+            grad_z0,
+            grad_args,
+            grad_terms,
+        )
+
+    def cond_fun(state):
+        ts_index = state[0]
+        return ts_index > 0
+
+    state = (
+        ts_final_index,
+        y1,
+        (original_solver_state, z1),
+        grad_y1,
+        grad_z1,
+        grad_args,
+        grad_terms,
+    )
+
+    state = eqxi.while_loop(cond_fun, grad_step, state, kind="lax")
+    _, _, _, grad_y0, grad_z0, grad_args, grad_terms = state
+
+    return (ω(grad_y0) + ω(grad_z0)).ω, grad_args, grad_terms
+
+
+class ReversibleAdjoint(AbstractAdjoint):
+    """
+    Backpropagate through [`diffrax.diffeqsolve`][] using the reversible solver
+    method.
+
+    This method automatically wraps the passed solver to create an algebraically
+    reversible version of that solver. In doing so, gradient calculation is exact
+    (up to floating point errors) and backpropagation becomes a linear in time $O(n)$
+    and constant in memory $O(1)$ algorithm in the number of steps $n$.
+
+    The reversible adjoint can be used when solving ODEs/CDEs/SDEs and is
+    compatible with any [`diffrax.AbstractSolver`][]. Adaptive step sizes are also
+    supported.
+
+    !!! note
+
+        This adjoint can be less numerically stable than
+        [`diffrax.RecursiveCheckpointAdjoint`][] and [`diffrax.DirectAdjoint`][].
+        Stability can be largely improved by using [double (64bit) precision](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision)
+        and [smaller/adaptive step sizes](https://docs.kidger.site/diffrax/api/stepsize_controller/).
+
+    ??? cite "References"
+
+        This algorithm was developed in:
+
+        ```bibtex
+        @article{mccallum2024efficient,
+            title={Efficient, Accurate and Stable Gradients for Neural ODEs},
+            author={McCallum, Sam and Foster, James},
+            journal={arXiv preprint arXiv:2410.11648},
+            year={2024}
+        }
+        ```
+
+        And built on previous work by:
+
+        ```bibtex
+        @article{kidger2021efficient,
+            title={Efficient and accurate gradients for neural sdes},
+            author={Kidger, Patrick and Foster, James and Li, Xuechen Chen and Lyons,
+                    Terry},
+            journal={Advances in Neural Information Processing Systems},
+            volume={34},
+            pages={18747--18761},
+            year={2021}
+        }
+        ```
+
+        ```bibtex
+        @article{zhuang2021mali,
+            title={Mali: A memory efficient and reverse accurate integrator for neural
+                    odes},
+            author={Zhuang, Juntang and Dvornek, Nicha C and Tatikonda, Sekhar and
+            Duncan, James S},
+            journal={arXiv preprint arXiv:2102.04668},
+            year={2021}
+        }
+        ```
+    """
+
+    l: float = 0.999
```

Comment:
Can we pick a more descriptive name for this parameter? E.g. when writing an optimizer then mathematically we may conventionally use λ but in code we would often use a variable with a name like `learning_rate`.